### DISCUSSION TITLE
[General] Issue with Mainnet RPC URL - Service Stops Responding after 24 Hours

### DISCUSSION THREAD

**Discussion by gauravl-tevaeralabs** [zkSync-Community-Hub/zksync-developers#262] [General]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

Our blockchain service stops receiving responses from the Mainnet RPC URL after approximately 24 hours of continuous operation. Restarting the service temporarily resolves the problem. The requests we make are primarily to the contract's read methods, such as retrieving token URLs, checking wallet balances, etc.

### What is the specific issue or error you're encountering?

No specific error messages or logs are available. The service simply stops receiving responses after a certain period.

### Can you share the error messages or logs you're receiving, if any?

No specific error messages or logs are available. The service simply stops receiving responses after a certain period.

### Have you made any recent changes to the contract before encountering this issue?

No recent changes have been made to the contract before encountering this issue.

### Are there any external libraries or contracts that your contract interacts with?

Our contract interacts with the ZKSync Mainnet RPC URL via zksync-web3 SDK and ethers sdk as an external source for blockchain.information.

### Can you provide the relevant portions of your contract code where the issue is occurring?

It occurs with any ERC721 standard read method such as balanceOf(), tokenURL(), etc.

### Have you tried to isolate the problem, and if so, what were the results?

We have attempted to isolate the problem, but without specific error messages, it's challenging to pinpoint the root cause.

### What steps have you already taken to try to resolve the issue?

- Restarting the service temporarily resolves the issue.
- Checked for the Mainnet RPC URL rate limits, but no conclusive information was found.
- Reviewed contract code for any potential issues, but no recent changes.

**Comment by MexicanAce** [comment]
When you mention that you stop receiving responses, does that mean the API calls are timing out?

**Reply by gauravl-tevaeralabs** [reply]
Yes. We're wondering if the zksync team has implemented any form of rate limiting or if there's a mechanism blocking our origin after a certain number of requests. Restarting the service resolves the issue temporarily, but we are looking for a permanent solution for the mainnet.

**Reply by MexicanAce** [reply]
Interesting, some additional questions:
- Are your connections via WebSockets?
- Are you also using a Public RPC or a Node Provider's RPC?
- Could you provide an example response when the timeout occurs? (Timestamp preferrable)

**Reply by gauravl-tevaeralabs** [reply]
Hi @MexicanAce. Please find the answers to your questions below:

**Q:** Are your connections via WebSockets?
**Ans:** No, we're not using WebSockets. We are using an HTTP URL.

**Q:** Are you also using a Public RPC or a Node Provider's RPC?
**Ans:** We are using a Public RPC URL, specifically https://mainnet.era.zksync.io/.

**Q:** Could you provide an example response when the timeout occurs? (Timestamp preferable)
**Ans:** Below are the errors we randomly encounter after a day:

_1) [1704793478241] ERROR: Missing revert data in call exception; upstream connect error or disconnect/reset before headers. Reset reason: remote connection failure, transport failure reason: delayed connect error: 111_

_2) [1704793536950] ERROR: Missing revert data in call exception; upstream connect error or disconnect/reset before headers. Reset reason: connection termination_

_3) [1704793417609] ERROR: Could not detect network (event="noNetwork", code=NETWORK_ERROR, version=providers/5.7.2)_

In addition to the above questions, one additional piece of information is that we create a JSON-RPC provider on service start and reuse it throughout the service's lifetime to enhance performance by avoiding the overhead of repeatedly establishing and closing connections. Could this be the issue here? As I mentioned earlier, this issue started in mid-December; before that, everything was working fine.

**Reply by gauravl-tevaeralabs** [reply]
Is there any update on this? It's critical for the mainnet.

**Reply by MexicanAce** [reply]
Hi @gauravl-tevaeralabs , apologies for the delayed response. A long-running provider/connection with the SDK to the RPC URL is something we utilize within our [Block Explorer](https://github.com/matter-labs/block-explorer) worker. The error for 111 could be associated with a request failing it's revalidation of the HTTPS with the upstream of the RPC URL after the 24 hour period.

Next steps: We should create a script that replicates the above scenario to replicate the errors. If you have this script already, please let us know/post it below and we'll try to replicate locally.

**Reply by gauravl-tevaeralabs** [reply]
Hi @MexicanAce, I don't have a script because we're only observing this issue in the production service, and even then, it's after a long period. It's difficult to reproduce locally!

**Reply by bxpana** [reply]
@gauravl-tevaeralabs wanted to circle back here and see if you guys are still having issues with this.

**Comment by loriyamazaki** [comment]
It sounds like your service is encountering some sort of resource exhaustion or connectivity issue after running continuously for about 24 hours. Since there are no specific error messages or logs available, it's challenging to identify the exact cause.

Here are some steps you can take to troubleshoot and resolve the issue:

Monitor Resource Usage: Keep an eye on system resource usage such as CPU, memory, and network bandwidth while the service is running. Check for any unusual spikes or patterns that may indicate resource exhaustion.

Check for Memory Leaks: Review your service code for any potential memory leaks that could gradually consume resources over time. Use memory profiling tools to identify and address any memory management issues.

Review Network Connectivity: Verify that your service has stable and uninterrupted network connectivity to the Mainnet RPC URL. Check for any network configuration issues, firewalls, or network congestion that may disrupt communication.

Handle Rate Limits: Ensure that your service adheres to any rate limits imposed by the Mainnet RPC URL provider. Excessive requests may result in throttling or temporary bans, leading to connectivity issues.

Implement Retry Mechanisms: Consider implementing retry mechanisms in your service to automatically retry failed requests to the Mainnet RPC URL. This can help mitigate transient network issues and improve overall reliability.

Monitor Logs and Metrics: Implement comprehensive logging and monitoring in your service to capture detailed information about its behavior and performance. This can help you identify patterns and trends that may indicate the root cause of the issue.

By systematically investigating these potential areas of concern and gathering more data through monitoring and logging, you should be able to diagnose and address the underlying issue causing your service to stop receiving responses after a certain period.

**Comment by mganlok** [comment]
Your service encountering a halt in receiving responses from the Mainnet RPC URL after approximately 24 hours of continuous operation suggests a potential issue with resource management or connectivity. Check for memory leaks, connection timeouts, or resource exhaustion in your service. Additionally, ensure proper error handling and logging to identify any underlying issues. Monitor system resource usage and network connectivity to pinpoint the cause of the interruption. #Blockchain #MainnetRPC #ServiceInterruption

**Comment by sonhoranemichelle** [comment]
It seems like your service may be experiencing issues related to resource exhaustion or connection stability over time. Consider implementing better error handling and logging mechanisms to gather more specific information when the issue occurs.

**Comment by segheamulchoidiaz** [comment]
The issue seems to be related to a blockchain service that stops receiving responses from the Mainnet RPC URL after approximately 24 hours of continuous operation. No specific error messages or logs are available, and the service simply stops receiving responses after a certain period. The requests made to the contract are primarily read methods such as retrieving token URLs and checking wallet balances. The contract interacts with the ZKSync Mainnet RPC URL via zksync-web3 SDK and ethers sdk as an external source for blockchain information. Restarting the service temporarily resolves the issue, but attempts to isolate the problem have been inconclusive.
